### PR TITLE
Dinas branch

### PR DESCRIPTION
**Pull request by dbash**
Working code for shared ResNet with weighted layers.

### CODE REVIEW

File: custom_resnet_exp.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+parser.add_argument("--kernel_reg",
+                    type=float,
+                    default=1e-5,
+                    help="kernel regularization parameter (default is 1e-5)")
+parser.add_argument("--lr",
+                    type=float,
+                    default=1e-3,
+                    help="learning rate (default is 0.01)")
+args = parser.parse_args()
+
```
Comment (azhmogin-google) [inline custom_resnet_exp.py]: I would suggest adding "main" function and breaking it up into a set of simpler functions.

File: custom_resnet_exp.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+# Normalize data.
+x_train = x_train.astype('float32') / 255
+x_test = x_test.astype('float32') / 255
+
+if args.pixel_mean:
+  x_train_mean = np.mean(x_train, axis=0)
+  x_train -= x_train_mean
+  x_test -= x_train_mean
+
+class LRTensorBoard(tf.keras.callbacks.TensorBoard):
```
Comment (azhmogin-google) [inline custom_resnet_exp.py]: Add a docstring here and to "on_epoch_end".

File: lib/custom_resnet.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
+import tensorflow as tf
+import lib.io_adapters as ioad
+import lib.res_block as block
+from tensorflow.keras import layers
+from tensorflow.keras import regularizers
+
+
+def resnet(input_shape, num_layers=20, num_classes=10, depth=40,
+             in_adapter="original", out_adapter="v1", tensor_size=8,
```
Comment (azhmogin-google) [inline lib/custom_resnet.py]: Wrong indentation.

File: lib/custom_resnet.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        `v1`, `v2`, `isometric` and `dethwise`.
+      kernel_regularizer: kernel regularization parameter.
+      dropout: dropout parameter (drop).
+    Returns
+      model (Model): Keras model instance
+    """
+
+    inputs = tf.keras.Input(shape=input_shape, name='input')
+    if in_adapter == 'original':
+        x = layers.Conv2D(depth, kernel_size=3, strides=1, padding='same',
```
Comment (azhmogin-google) [inline lib/custom_resnet.py]: Wrong indentation here and later.

File: lib/custom_resnet.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    else:
+        raise ValueError("Given input adapter type is not supported.")
+
+    # Instantiate the stack of residual units
+    for stack in range(num_layers):
+       x = block.ResBlock(kernel_size=3, expansion_factor=6,
+                          activation='swish',
+                          kernel_reg=kernel_regularizer)(x)
+
+    if out_adapter == 'v1':
```
Comment (azhmogin-google) [inline lib/custom_resnet.py]: Output adapter builder can be moved into a separate function. (Maybe other parts too?)

File: lib/custom_resnet.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        in_shape = (tensor_size, tensor_size, depth)
+        out_adapter = create_output_adapter_isometric(
+            in_shape, filters=[128, 256], pool_size=8, activation='relu',
+            dropout=dropout)
+        x = out_adapter(x)
+    elif out_adapter == 'dethwise':
+        t_shape = input_shape[0] if in_adapter == 'original' else tensor_size
+        in_shape = (tensor_size, tensor_size, depth)
+        out_adapter = create_output_adapter(
+            in_shape,  block_size=2, pool_stride=None,
```
Comment (azhmogin-google) [inline lib/custom_resnet.py]:
```suggestion
            in_shape, block_size=2, pool_stride=None,
```

Make sure pylint is catching these and wrong indentation.

File: lib/shared_resnet.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+                      kernel_regularizer=regularizers.l2(1e-4)
+                      )(inputs)
+    x = layers.BatchNormalization()(x)
+    x = layers.Activation('relu')(x)
+  elif in_adapter == 'space2depth':
+    in_adapter = ioad.create_input_adapter(
+        input_shape, size=tensor_size, depth=depth, activation='relu')
+    x = in_adapter(inputs)
+  elif in_adapter == 'strided':
+    k = max(input_shape[0]//tensor_size, input_shape[0]//tensor_size)
```
Comment (azhmogin-google) [inline lib/shared_resnet.py]:
```suggestion
    k = max(input_shape[0] // tensor_size, input_shape[0] // tensor_size)
```

File: lib/shared_resnet.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  xi_initializer = tf.random_uniform_initializer(minval=0.0, maxval=1.0)
+
+  # Instantiate the stack of residual units
+  for _ in range(num_layers):
+    # mixture weights for this layer
+    xi = wb.MixtureWeight(num_templates=num_templates,
+                          initializer=xi_initializer)(x)
+    x = w_res_block([x, xi])
+
+  if out_adapter == 'v1':
```
Comment (azhmogin-google) [inline lib/shared_resnet.py]: Move out to a separate function / reuse this function across implementations when possible.

File: tests/weighted_layers_test.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    input_shape = (256, 16, 16, self.input_channels)
+    inputs = tf.random.normal(input_shape)
+    bn = WeightedBatchNormalizationSeparate(num_templates=self.num_templates,
+                                            gamma_initializer=self.gamma_init,
+                                            beta_initializer=self.beta_init)
+    outputs = bn([inputs, self.xi(None)], training=True)
+    self.assertAllEqual(input_shape, outputs.shape)
+
+  def test_output_moments(self):
+    """checks if the moments of the output tensor match to the value of mixture
```
Comment (azhmogin-google) [inline tests/weighted_layers_test.py]: Great test! Side note: I would try to shrink the function docstring so that it fits in a single line.

File: lib/weighted_resblock.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+    x = self.conv2([x, mix_weights])
+    x = self.bn2([x, mix_weights], training)
+    x = self.activation(x)
+
+    x = self.conv3([x, mix_weights])
+    x = self.bn3([x, mix_weights], training)
+    x += layer_input
+    return x
+
+  def set_activation_fn(self, activation):
```
Comment (azhmogin-google) [inline lib/weighted_resblock.py]: I would call it differently. It does not really set the activation function, but maps the name of the activation to the actual function.

File: lib/weighted_resblock.py [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+    self.expansion_factor = expansion_factor
+    self.activation = self.set_activation_fn(activation)
+    self.kernel_size = kernel_size
+    self.template = ResBlockTemplate() if template is None else template
+    self.num_templates = num_templates
+    self.kernel_reg = kernel_reg
+
+  def build(self, input_shape):
+    input_channel = input_shape[0][-1]
+    self.expanded_channel = input_channel*self.expansion_factor
```
Comment (azhmogin-google) [inline lib/weighted_resblock.py]:
```suggestion
    self.expanded_channel = input_channel * self.expansion_factor
```

File: lib/weighted_resblock.py [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+        num_templates=self.num_templates, gamma_initializer=gamma,
+        beta_initializer=beta, moving_mean_initializer=mean,
+        moving_variance_initializer=var)
+
+    depthwise_kernel_init, bias_init = self.template.get_depthwise_template()
+    self.conv2 = wl.WeightedDepthwiseConv2D(
+        kernel_size=self.kernel_size, strides=(1, 1), padding='same',
+        num_templates=self.num_templates,
+        depthwise_initializer=depthwise_kernel_init,
+        #depthwise_regularizer=regularizers.l2(1e-4),
```
Comment (azhmogin-google) [inline lib/weighted_resblock.py]: Remove?

File: lib/weighted_layers_v2.py [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+    layer_inputs = inputs[0]
+    mix_weights = inputs[1]
+    self.assign_mixture_value(name="template_beta", mixture_weights=mix_weights)
+    self.assign_mixture_value(name="template_gamma", mixture_weights=mix_weights)
+    norm_input = BatchNormalization.call(self, layer_inputs, training)
+
+    input_shape = layer_inputs.shape
+    reduction_axes = [i for i in range(len(input_shape))
+                      if i not in self.axis]
+    mean, var = nn.moments(norm_input, reduction_axes, keep_dims=True)
```
Comment (azhmogin-google) [inline lib/weighted_layers_v2.py:197]: I assume in the training mode, we really expect mean to be 0 and var to be 1?

File: custom_resnet_exp.py [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+  def __init__(self, log_dir, **kwargs):
+    super().__init__(log_dir=log_dir, **kwargs)
+
+  def on_epoch_end(self, epoch, logs=None):
+    logs = logs or {}
+    logs.update({'lr': K.eval(self.model.optimizer.lr)})
+    super().on_epoch_end(epoch, logs)
+
+def lr_schedule(epoch):
```
Comment (azhmogin-google) [inline custom_resnet_exp.py]: Experiments share some code (like this function). Makes sense to move it out to shared "util.py" or something along these lines.

**Review by azhmogin-google — APPROVED** [review]
Overall great implementation. There are a few cosmetic suggestions, but the core of the implementation seems to be correct and covered by reasonable tests. Great work!

File: custom_resnet_exp.py (hunk 1 above)
Comment (dbash) [reply custom_resnet_exp.py]: Good point, will do!

File: lib/weighted_layers_v2.py (hunk 13 above)
Comment (dbash) [reply lib/weighted_layers_v2.py:197]: Yes, but it might be a little bit off these values. The description of BN layer says it will normalize to something close to zero mean and variance 1. I have tried to compare with constants at first, and the corresponding tests crashed.
